### PR TITLE
frame: Enable GenesisConfig in no_std

### PR DESCRIPTION
**Pull request by michalkucharczyk**
This change enables `GenesisConfig` compilation in `no_std` environment.

It also enables `Default` implementation for `GenesisConfig` which will be required for no_std in no-native
runtime world. It must be possible to instantiate default GenesisConfig for pallets and runtime.

Step towards: paritytech/polkadot-sdk#25

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2807906

**Comment by michalkucharczyk** [comment]
bot format

**Comment by michalkucharczyk** [comment]
bot fmt

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2808154

**Review by bkchr — APPROVED** [review]

**Comment by bkchr** [comment]
@michalkucharczyk please change the title and add some more description to the pr. This can break downstream code as `GenesisConfig` starts to compile for `no_std` now, which wasn't the case before. So, please mention this :)

File: frame/alliance/src/lib.rs [hunk 1]
```diff
@@ -407,7 +407,6 @@ pub mod pallet {
 		pub phantom: PhantomData<(T, I)>,
 	}
 
-	#[cfg(feature = "std")]
```
Comment (ggwpez) [inline frame/alliance/src/lib.rs:410]: Should be possible to replace the impl here with `derive(DefaultNoBound)`.

File: frame/assets/src/lib.rs [hunk 2]
```diff
@@ -375,7 +375,6 @@ pub mod pallet {
 		pub accounts: Vec<(T::AssetId, T::AccountId, T::Balance)>,
 	}
 
-	#[cfg(feature = "std")]
```
Comment (ggwpez) [inline frame/assets/src/lib.rs:378]: dito and more below.

**Review by ggwpez — COMMENTED** [review]
Eventually we could add a `derive(DefaultNoBound)` to the `genesis_config` macro, but that would be a breaking change.

**Comment by michalkucharczyk** [comment]
> @michalkucharczyk please change the title and add some more description to the pr. This can break downstream code as `GenesisConfig` starts to compile for `no_std` now, which wasn't the case before. So, please mention this :)

Initially it was only about `Default` (facepalm).
Done.

**Comment by michalkucharczyk** [comment]
> Do we expect to put the conditional compilation of the genesis builders into the WASM binary in `construct_runtime`? Or should every pallet also contain a feature gate for WASM genesis?

The current plan is to not add any feature (or conditional compilation) for genesis config in pallets.

Support for `GenesisConfig` building in runtime (the new runtime API will be introduced for that) will be guarded with new feature. We expect linker to do the job and strip unnecessary `GenesisConfig` related staff if the feature is not enabled.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2814036

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2814037

File: frame/alliance/src/lib.rs (hunk 1 above)
Comment (michalkucharczyk) [reply frame/alliance/src/lib.rs:410]: done in: https://github.com/paritytech/substrate/pull/14108/commits/8a21eb158b0c03f2213fd8dfcafcaa9d0710397b

File: frame/assets/src/lib.rs (hunk 2 above)
Comment (michalkucharczyk) [reply frame/assets/src/lib.rs:378]: all done in: https://github.com/paritytech/substrate/pull/14108/commits/8a21eb158b0c03f2213fd8dfcafcaa9d0710397b

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2814597

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2814699

File: frame/support/procedural/src/pallet/expand/genesis_config.rs [hunk 3]
```diff
@@ -126,7 +135,7 @@ pub fn expand_genesis_config(def: &mut Def) -> proc_macro2::TokenStream {
 						stringify!($pallet_name),
 						"` does not have the std feature enabled, this will cause the `",
 						$pallet_path,
-						"::GenesisConfig` type to be undefined."
+						"::GenesisConfig` type to not implement serde traits."
```
Comment (michalkucharczyk) [inline frame/support/procedural/src/pallet/expand/genesis_config.rs:138]: When std feature is not enabled for pallet, the GenesisConfig will be defined, but `serde::{Serialize,Deserialize}` traits will not be implemented.
    
I've decided to change this compiler error to indicate the reason of latter errors occuring during compilation.
    
This is actually temporary and serde traits will be enabled with together with `serde` support in frame.

**Comment by michalkucharczyk** [comment]
bot rebase

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/genesisconfig-in-wasm-runtime-a-step-towards-a-non-native-future/2887/1

**Review by sam0x17 — APPROVED** [review]

**Comment by michalkucharczyk** [comment]
bot rebase

File: frame/support/procedural/src/pallet/expand/genesis_config.rs [hunk 4]
```diff
@@ -94,15 +94,24 @@ pub fn expand_genesis_config(def: &mut Def) -> proc_macro2::TokenStream {
...
+				syn::parse_quote!( #[cfg_attr(feature = "std", serde(deny_unknown_fields))] ),
+			);
+			attrs.push(
+				syn::parse_quote!( #[cfg_attr(feature = "std", serde(bound(serialize = "")))] ),
+			);
+			attrs.push(
+				syn::parse_quote!( #[cfg_attr(feature = "std", serde(bound(deserialize = "")))] ),
+			);
+			attrs.push(
+				syn::parse_quote!( #[cfg_attr(feature = "std", serde(crate = #serde_crate))] ),
```
Comment (ggwpez) [inline frame/support/procedural/src/pallet/expand/genesis_config.rs:113]: Maybe you can push them all in one `quote` call?

**Review by ggwpez — APPROVED** [review]

File: frame/support/procedural/src/pallet/expand/genesis_config.rs (hunk 4 above)
Comment (KiChjang) [reply frame/support/procedural/src/pallet/expand/genesis_config.rs:113]: Indeed, this is possible through binding a new `TokenStream` called `serde_attr_arg` or something similar, and use the `quote!` macro for it.

Comment (bkchr) [reply frame/support/procedural/src/pallet/expand/genesis_config.rs:113]: this is `parse_quote!` and will not work.

**Review by shawntabrizi — APPROVED** [review]
awesome

**Comment by michalkucharczyk** [comment]
bot merge

**Pull request merged**